### ISSUE TITLE
Adding `observer` affects original component

### ISSUE THREAD

**Issue by sergei-startsev** [mobxjs/mobx-react#697]
**What is the current behavior?**
Adding `observer` wrapper affects original `render` method. There's some regression after upgrading `mobx-react`  to version **6** caused by affecting the original (**unwrapped**) component.

```jsx
class Foo extends React.Component {
  render() {
    //...
  }
}

const observableFoo = observer(Foo);
```
`observableFoo.render` as well as original `Foo.render` is overridden by adding  
```jsx
<Observer>
    <Component />
</Observer>
```
Since `observer` isn't pure, there's no possibility to reuse a component without mobx.

Here's a repository with repro steps: https://github.com/sergei-startsev/mobx-react-observer

**What is the expected behavior?**

There's an expectation that adding `observer` wrapper should not affect the original component. E.g.  adding `observer` wrapper in the default export below should not affect named `Foo` export: 
```jsx
export class Foo extends React.Component {
  // ...
  render() {
  }
}
export default inject('store')(observer(Foo));


// Foo shouldn't be wrapped
import { Foo } from './Foo.js';
```

**Comment by FredyC** [comment]
I don't follow what's the issue. If you use exported `Foo`, it's unaffected by observer and as such can be reused without knowing anything about MobX. If you use default export in tests, it will have Observer in there, otherwise, it won't be there. I wouldn't consider it impure. Using the same input gives the same output.

Can you elaborate more please?

**Comment by urugator** [comment]
Seems to be done deliberately, check this PR
https://github.com/mobxjs/mobx-react/commit/d3820da16e45bf2b97695bb465605cb2d2faea23#diff-3035dd5a37335cc50db60f91ad838797
Maybe there is some reasoning in the original PR:
https://github.com/mobxjs/mobx-react/pull/523
And it's also mentioned in this comment (last sentence):
https://github.com/mobxjs/mobx-react/issues/588#issuecomment-436228990

**Comment by sergei-startsev** [comment]
@FredyC 
>If you use exported Foo, it's unaffected by observer and as such can be reused without knowing anything about MobX.

It is affected, take a look at `Foo > shallow` test from [repro steps](https://github.com/sergei-startsev/mobx-react-observer). It uses named unwrapped `Foo` export, however its snapshoot contains Observer component:
```diff
- <div>
-   Test
- </div>
+ <Observer>
+   <Component />
+ </Observer>
```

**Comment by urugator** [comment]
Is this a viable workaround?
`export default inject('store')(observer(class ObserverFoo extends Foo {}));`

**Comment by FredyC** [comment]
@sergei-startsev I am looking at a snapshot file and I don't see any Observer there. That's why I am confused. 

https://github.com/sergei-startsev/mobx-react-observer/blob/master/__snapshots__/Foo.test.jsx.snap#L9

**Comment by sergei-startsev** [comment]
@urugator 
Yes, it works, but it doesn't look like a workaround though. I suppose the original `render` method shouldn't be overriden:
https://github.com/mobxjs/mobx-react/blob/4904b624f46be2e14d3a3d1017c1fa9bcb334f02/src/observer.js#L140

It seems to me that a new inherited class has to be created instead of changing prototype of the original:
https://github.com/mobxjs/mobx-react/blob/4904b624f46be2e14d3a3d1017c1fa9bcb334f02/src/observer.js#L126

**Comment by sergei-startsev** [comment]
@FredyC 
Take a look at `mobx-react 6` step: https://github.com/sergei-startsev/mobx-react-observer#mobx-react-6

**Comment by FredyC** [comment]
I see your step, but writing something in readme does not make it true 😆 Why test snapshot is showing something else?

**Comment by urugator** [comment]
> It seems to me that a new inherited class has to be created instead of changing prototype of the original:

Have you checked my first comment?
There was a PR doing what you're suggesting, but the PR was quickly reverted after a month or two.

**Comment by sergei-startsev** [comment]
@FredyC I've added both examples with `mobx-react` 5 and 6 to demo different results. See the issue description:
>There's some regression after upgrading mobx-react to version 6 caused by affecting the original (unwrapped) component.

Just run steps from `mobx-react` 6 to see the diff.

**Comment by FredyC** [comment]
It would be much more helpful if the snapshot would be showing an actual problem.

Either way, check the discussion on why the change you are suggesting was reverted. If you have idea how to make that work, PR is most welcome.

https://github.com/mobxjs/mobx-react/pull/523#issuecomment-423180044

**Comment by mweststrate** [comment]
See also #703, this behavior might need to be reverted anyway (not that it really does make a fundamental difference, v5 is also patching classes, but at least the behavior would be the same)

That being said, this problem is unavoidable from an implementation level, as we have to reach into some of the React internals. If you want a pure declaration: don't use `observer`, but use `Observer` instead. That is the "proper" composition abstraction React, but not without it's own downsides, such as making it hard to coordinate life cycle events between the "outer" component and reactive rendering in the "inner" component.

**Comment by mweststrate** [comment]
Fixed in 6.1.0
